File: Projects/CPP-VNEXT/src/Animals/Fox.cpp
#include <optional>
#include "Fox.h"
#include "../World.h"

Fox::Fox(Position&& position) : Animal(3, 7, 'F', std::move(position)) {

}

const std::string Fox::Type = "Fox";

std::string Fox::GetType() const {
    return Fox::Type;
}

Organism* Fox::GetNewOfType(Position&& position) {
    return new Fox(std::move(position));
}

void Fox::HandleAction(World& world) {
    std::optional<Position> newPosition = world.GetNearbyPosition(this->GetPosition(), 1, true);
    if (newPosition && world.Move(this->GetPosition(), *newPosition)) {
        this->SetPosition(std::move(*newPosition));
    }
}

File: Projects/Java/src/main/java/com/nieradko/worldsim/core/animals/Fox.java
package com.nieradko.worldsim.core.animals;

import com.nieradko.worldsim.core.IActionContext;
import com.nieradko.worldsim.core.Position;

public class Fox extends Animal {
    public Fox(Position position) {
        super(3, 7, position);
    }

    @Override
    protected void handleAction(IActionContext context) {
        context.getRandomNearbyPosition(getPosition(), true)
                .ifPresent(value -> context.move(this, value));
    }
}
